### ISSUE TITLE
Improve code coverage with more RAR decodes

### ISSUE THREAD

**Issue by simsong** [simsong/bulk_extractor#394]
Looks like we can significantly increase coverage by having more RAR decompression examples.

**Comment by MasterDFOR** [comment]
Hi Simsong, 

I was working on this project and created datasets to see which RAR files cannot be detected by Bulk Extractor. The RAR files created in any compression mode in RAR5 via WinRAR cannot be detected by Bulk Extractor. All RAR files created in RAR4 or in any older versions via WinRAR can be detected. Hope this finding can contribute for further research.

**Comment by simsong** [comment]
Hi. Thank you so much. It would be great to have this dataset. Where is it? Can we add it to the repo?

**Comment by MasterDFOR** [comment]
Yes, I add add it. Where is the repo you mentioned btw? The dataset are a few dd images which contains RAR files that I created via different compression levels (fast, normal, good, etc) and compression modes (RAR4 or RAR5).

**Comment by simsong** [comment]
How large are the files? If they are under 5MB, please attach them to this ticket.

**Comment by simsong-codex** [comment]
## RAR implementation audit and proposed work sequence

This comment memorializes the repository and design review requested by the project owner.

### What is currently in the repository

- `src/rar/version.hpp` identifies the vendored decoder as UnRAR 4.10 from 2012.
- `src/rar/` contains 78 C++ source/header files and about 15,500 lines.
- `scan_rar` recognizes the seven-byte RAR 4.x signature (`52 61 72 21 1a 07 00`). RAR 5.0 uses a distinct eight-byte signature (`52 61 72 21 1a 07 01 00`) and different variable-length headers, so the current scanner cannot identify or decode RAR5.
- RAR4 is a real format generation: it is the final generation of the legacy RAR family, distinct from RAR5.
- `tests/Data/rar_samples.tar` already contains samples produced by RAR 2.0, 2.06, 2.5, 2.9, 3.11, 3.2, 3.93, 4.0, and 4.11, including solid, SFX, locked, and password variants.
- The repository also contains `jpegs.rar`, `jpegs.rar-encrypted`, `jpegs.rar5`, and `jpegs.rar5-encrypted`.
- No RAR1 fixture was found.
- Existing automated coverage is much narrower than this corpus. The phase-one test asserts only one recursively recovered JPEG, while `tests/data_check.txt` records three carved RAR members. Neither verifies all four members or provides a cross-version acceptance matrix.

### Reliability and maintainability findings

- The decoder retains shared mutable state, including global `ErrHandler` state and static file bookkeeping. It must not be assumed thread-safe.
- `scan_rar` drives the old command-line-oriented `CommandData`/`CmdExtract` stack to perform an in-memory decode. That pulls command parsing, filesystem extraction, path handling, messages, resources, and other likely-unneeded code into the build.
- Apparent dead code should not be deleted solely by filename. First replace the command-line entry point with a narrow memory-to-memory decoder interface, generate a linker map/reachability report, and remove one proven-unused cluster at a time.
- Expanded sizes are derived from untrusted archive headers, and the current suspicious-size ceiling is measured in terabytes. Any decoder work must add explicit per-member output, aggregate output, expansion-ratio, recursion-depth, and time budgets before increasing format coverage.
- Metadata/header carving and decompression should become separate layers. The metadata scanner can continue identifying embedded RAR structures, while decompression should start only from a validated archive boundary and operate on bounded input.

### Proposed sequence

1. **Build a versioned fixture matrix.**
   - Inventory every existing fixture with provenance, format generation, archive features, expected members, sizes, and hashes.
   - Exercise RAR2, RAR3, RAR4, and RAR5 in normal CI.
   - Locate or generate a legally redistributable RAR1 fixture with documented provenance.
   - Include stored/compressed, solid, encrypted-header, split/truncated, SFX, malformed, and embedded-offset cases where applicable.
   - Run the same corpus concurrently to expose shared-state problems.

2. **Resolve #212 against the legacy decoder.**
   - Explain why all four JPEG members are not recovered.
   - Make the existing RAR4 fixture produce all expected members before changing decoder implementations.
   - Record exact member hashes and reject partial/truncated output unless explicitly classified as such.

3. **Introduce a bounded decoder interface and remove dead code.**
   - Accept a bounded input span and output callbacks/sbufs.
   - Enforce resource budgets centrally.
   - Eliminate global mutable state or temporarily serialize the old decoder until reentrancy is demonstrated.
   - Use linker reachability evidence and the complete corpus before deleting old UnRAR source clusters.

4. **Evaluate RAR5 under #223.**
   - Do not implement the proprietary decompressor directly from the format specification.
   - Prototype libarchive first because it provides separate RAR and RAR5 readers, memory/custom I/O, and a maintained BSD-licensed implementation.
   - Validate libarchive against the complete legacy and RAR5 corpus, especially solid archives, damaged input, embedded archives, concurrency, bounded expansion, and extraction completeness.
   - Verify that Ubuntu/MinGW still produces a statically distributed executable with no new external DLLs.
   - If libarchive does not preserve required forensic behavior, evaluate adapting current official UnRAR behind the same bounded interface.

### Acceptance gates

- All expected members and hashes match for the supported fixture matrix.
- Malformed input cannot allocate unbounded memory, hang indefinitely, write outside the output directory, or terminate the process.
- Concurrent scans are deterministic and race-free.
- The scanner preserves forensic offsets and clearly distinguishes metadata identification, successful decompression, encrypted/unsupported data, and damaged archives.
- Linux, macOS, and Ubuntu/MinGW CI pass, and the Windows artifact remains self-contained.

References:

- RARLAB RAR 5.0 format description: https://www.rarlab.com/technote.htm
- libarchive format support: https://github.com/libarchive/libarchive
- Legacy extraction defect: #212
- RAR4/RAR5 support request: #223

_Analysis prepared by @simsong-codex at the repository owner's request. No RAR implementation was started by this update._
